File: 20-actors/todoke/methods/test_last_mile.py
"""Tests for methods/last_mile.py — sequencer correctness, G7 envelope, courier sizing.

The PARITY contract: this Python sequencer and the Rust `todoke-route` crate must return
the SAME visiting order on the shared fixtures. The Rust side pins
[0, 4, 2, 3, 1] / length 30 in `route/src/lib.rs::sequences_collinear_stops_in_order`;
``test_parity_collinear_matches_rust`` pins the identical result here.
"""

import pytest

from last_mile import (
    EnvelopeViolation,
    Stop,
    courier_freed_hours,
    displacement_cohort_size,
    plan_last_mile,
)

_COLLINEAR = [
    Stop(0, 0.0, 0.0, "sidewalk"),
    Stop(1, 30.0, 0.0, "doorpath"),
    Stop(2, 10.0, 0.0, "sidewalk"),
    Stop(3, 20.0, 0.0, "doorpath"),
    Stop(4, 5.0, 0.0, "crosswalk"),
]


def test_parity_collinear_matches_rust():
    order, length = plan_last_mile(_COLLINEAR, sae_level=4, commanded_mps=1.0)
    assert order == [0, 4, 2, 3, 1]          # identical to the Rust crate fixture
    assert abs(length - 30.0) < 1e-6


def test_two_opt_removes_crossing_on_square():
    sq = [
        Stop(0, 0.0, 0.0, "sidewalk"),
        Stop(1, 0.0, 10.0, "sidewalk"),
        Stop(2, 10.0, 10.0, "sidewalk"),
        Stop(3, 10.0, 0.0, "sidewalk"),
    ]
    _, length = plan_last_mile(sq, commanded_mps=1.5)
    assert length <= 30.0 + 1e-6


def test_g7_refuses_sae_5():
    with pytest.raises(EnvelopeViolation, match="exceeds ceiling"):
        plan_last_mile(_COLLINEAR, sae_level=5, commanded_mps=1.0)


def test_g7_refuses_road_zone():
    stops = _COLLINEAR + [Stop(9, 40.0, 0.0, "road")]
    with pytest.raises(EnvelopeViolation, match="outside todoke ODD"):
        plan_last_mile(stops, commanded_mps=1.0)


def test_g7_refuses_speed_over_cap():
    with pytest.raises(EnvelopeViolation, match="exceeds"):
        plan_last_mile(_COLLINEAR, commanded_mps=3.0)


def test_empty_refused():
    with pytest.raises(EnvelopeViolation, match="no stops"):
        plan_last_mile([])


def test_courier_sizing_is_positive():
    assert courier_freed_hours(1.0e7, 2200, 0.3) > 0
    assert displacement_cohort_size(1.0e7, 0.3) == 3_000_000

File: 20-actors/todoke/methods/test_last_mile.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_last_mile.py + the ainori reuse-parity check.
(ns todoke.methods.test-last-mile
  "Tests for methods/last_mile.clj — sequencer correctness, G7 envelope, courier sizing, and the
  cross-actor reuse parity: ainori's clj sequencer returns the SAME visiting order as todoke's
  plan-last-mile (the ADR-2606071500 'one engine, not a fork' claim, now both in clj).

  Run:  bb --classpath 20-actors 20-actors/todoke/methods/test_last_mile.clj"
  (:require [todoke.methods.last-mile :as lm]
            [ainori.methods.pooled-route :as pr]
            [clojure.test :refer [deftest is run-tests]]))

(defn- collinear []
  [(lm/stop 0 0.0 0.0 "sidewalk") (lm/stop 1 30.0 0.0 "doorpath") (lm/stop 2 10.0 0.0 "sidewalk")
   (lm/stop 3 20.0 0.0 "doorpath") (lm/stop 4 5.0 0.0 "crosswalk")])

(deftest parity-collinear-matches-rust
  (let [[order length] (lm/plan-last-mile (collinear) :sae-level 4 :commanded-mps 1.0)]
    (is (= order [0 4 2 3 1]))                 ; identical to the Rust crate + Python fixture
    (is (< (Math/abs (- length 30.0)) 1e-6))))

(deftest two-opt-removes-crossing-on-square
  (let [sq [(lm/stop 0 0.0 0.0 "sidewalk") (lm/stop 1 0.0 10.0 "sidewalk")
            (lm/stop 2 10.0 10.0 "sidewalk") (lm/stop 3 10.0 0.0 "sidewalk")]
        [_ length] (lm/plan-last-mile sq :commanded-mps 1.5)]
    (is (<= length (+ 30.0 1e-6)))))

(deftest g7-refuses-sae-5
  (is (thrown? clojure.lang.ExceptionInfo (lm/plan-last-mile (collinear) :sae-level 5 :commanded-mps 1.0))))

(deftest g7-refuses-road-zone
  (is (thrown? clojure.lang.ExceptionInfo
               (lm/plan-last-mile (conj (collinear) (lm/stop 9 40.0 0.0 "road")) :commanded-mps 1.0))))

(deftest g7-refuses-speed-over-cap
  (is (thrown? clojure.lang.ExceptionInfo (lm/plan-last-mile (collinear) :commanded-mps 3.0))))

(deftest empty-refused
  (is (thrown? clojure.lang.ExceptionInfo (lm/plan-last-mile []))))

(deftest refusals-are-envelope-violations
  (is (try (lm/plan-last-mile []) false
           (catch clojure.lang.ExceptionInfo e (lm/envelope-violation? e)))))

(deftest courier-sizing-is-positive
  (is (> (lm/courier-freed-hours 1.0e7 2200 0.3) 0))
  (is (= (lm/displacement-cohort-size 1.0e7 0.3) 3000000)))

;; ── cross-actor reuse parity: ainori reuses todoke's sequencer, not a fork ──
(deftest ainori-sequencer-matches-todoke
  (let [fixture [{:id 0 :x 0.0 :y 0.0 :zone "sidewalk"} {:id 1 :x 3.0 :y 0.0 :zone "sidewalk"}
                 {:id 2 :x 3.0 :y 3.0 :zone "sidewalk"} {:id 3 :x 0.0 :y 3.0 :zone "sidewalk"}
                 {:id 4 :x 1.0 :y 1.0 :zone "sidewalk"}]
        [order-a len-a] (pr/sequence-stops fixture)
        [order-t len-t] (lm/plan-last-mile (mapv #(lm/stop (:id %) (:x %) (:y %) (:zone %)) fixture)
                                           :sae-level 4 :commanded-mps 1.5)]
    (is (= order-a order-t) "ainori clj sequencer == todoke clj plan-last-mile (one engine, not a fork)")
    (is (< (Math/abs (- len-a len-t)) 1e-9))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'todoke.methods.test-last-mile)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
